aptos-move/framework/aptos-framework/sources/configs/Stake.move

```diff
@@ -276,167 +413,313 @@ module AptosFramework::Stake {
 
     /// Update individual validator's stake pool
     /// 1. distribute rewards to active/pending_inactive delegations
-    /// 2. process pending_active, pending_inactive correspondingly
+    /// 2. process pending_active, pending_inactive delegations correspondingly
     /// This function shouldn't abort.
-    fun update_stake_pool(addr: address) acquires StakePool {
-        let stake_pool = borrow_global_mut<StakePool>(addr);
-        distribute_reward( &mut stake_pool.active);
-        distribute_reward( &mut stake_pool.pending_inactive);
+    fun update_stake_pool(validator_address: address) acquires Delegation, StakePool {
+        let stake_pool = borrow_global_mut<StakePool>(validator_address);
+        let pending_inactive_pool = &mut stake_pool.pending_inactive;
+        distribute_reward(&mut stake_pool.active);
+        distribute_reward(&mut pending_inactive_pool.pool);
         // move pending_active to active
-        IterableTable::append(&mut stake_pool.active, &mut stake_pool.pending_active);
+        process_pending_pool(&mut stake_pool.active, &mut stake_pool.pending_active, DELEGATION_STATE_ACTIVE);
         // move pending_inactive to inactive
-        IterableTable::append(&mut stake_pool.inactive, &mut stake_pool.pending_inactive);
+        process_pending_pool(&mut stake_pool.inactive, pending_inactive_pool, DELEGATION_STATE_INACTIVE);
     }
 
-    /// Mint the reward and add to the delegation based on some formula
-    fun distribute_reward(v: &mut IterableTable<address, Delegation>) {
-        let key = IterableTable::head_key(v);
-        while (Option::is_some(&key)) {
-            let (delegation, _, next) = IterableTable::borrow_iter_mut(v, Option::borrow(&key));
-            let reward = Coin::zero<TestCoin>(); // mint some coins based on delegation, timestamp, maybe also total stakes
-            Coin::merge<TestCoin>(&mut delegation.rewards, reward);
-            key = next;
+    /// For each delegator, withdraw coins from `pending_pool` and deposit in `pool`.
+    fun process_pending_pool(
+        pool: &mut CoinPool,
+        pending_pool: &mut PendingCoinPool,
+        new_delegation_state: u64,
+    ) acquires Delegation {
+        let delegator_addresses = &mut pending_pool.delegator_addresses;
+        let i = 0;
+        let len = Vector::length(delegator_addresses);
+        while (i < len) {
+            let delegator_address = *Vector::borrow(delegator_addresses, i);
+            let delegation = borrow_global_mut<Delegation>(delegator_address);
+            let coins = extract_from_pool(&mut pending_pool.pool, delegation.shares);
+            delegation.shares = buy_shares_from_pool(pool, coins);
+            delegation.state = new_delegation_state;
+
+            i = i + 1;
         };
+        // There might be some coin residual (dust) left in the pending_pool but this should not affect operations.
```

Comment:
it's not dust? the shares in pending_pool is very different than the actual share they're going to get from the actual pool

Reply:
this probably shouldn't be a CoinPool, it's simpler to have a vector<Coin>, vector<address>

Reply:
The dust comment is probably left over from a different way of doing things. Having CoinPool here makes thing simpler since I can use the standard extract_from_pool/buy_shares_from_pool combo without having to do separate shares calculation here for each delegator.